### DISCUSSION TITLE
Migrating archive to sql issues

### DISCUSSION THREAD

**Discussion by Saphar** [processone/ejabberd#4073] [Questions] [answered]
Hello, I am trying to follow the tutorial to migrate mnesia to sql:
https://www.process-one.net/blog/install-and-configure-mariadb-with-ejabberd/

I am experiencing problems with the export of the DB:
```
ejabberdctl export2sql HOST /tmp/mnesia.sql
```

I have this installation on debian from bullseye-backports, the node is running:
```
ejabberdctl status
The node ejabberd@localhost is started with status: started
ejabberd 23.01-1~bpo11+1 is running in that node
```

I tried first:
```
ejabberdctl export2sql mydomain.ltd /tmp/mnesia.sql
```
gives the error:
```
Failed RPC connection to the node ejabberd@localhost: nodedown
```

Then I tried:
```
ejabberdctl export2sql localhost /tmp/mnesia.sql
```
Giving the error:
```
Unhandled exception occurred executing the command:
** exception error: {module_not_loaded,mod_muc,<<"localhost">>}
   in function  gen_mod:get_module_opts/2 (gen_mod.erl, line 338)
   in call from gen_mod:get_module_opt_hosts/2 (gen_mod.erl, line 323)
   in call from ejd2sql:export/3 (ejd2sql.erl, line 77)
   in call from lists:foreach/2 (lists.erl, line 1342)
   in call from ejd2sql:export/2 (ejd2sql.erl, line 70)
   in call from ejabberd_ctl:call_command/4 (ejabberd_ctl.erl, line 334)
   in call from ejabberd_ctl:try_call_command/4 (ejabberd_ctl.erl, line 295)
   in call from ejabberd_ctl:process2/4 (ejabberd_ctl.erl, line 235)
```

In the logs I see the following error:
```
2023-08-04 04:36:46.488570-04:00 [warning] <0.618.0>@ejabberd_sql:handle_reconnect/2:479 odbc connection failed:
** Reason: {"IM002",0,
            "[unixODBC][Driver Manager]Data source name not found, and no default driver specified Connection to database failed."}
** Retry after: 30 seconds
```

I do not have any odbc configuration in the config file, I have nothing configured about databases in the config file. But I need to migrate because my archive is over 2GB. Can anyone help me?

**Comment by prefiks** [comment]
Is your host really named localhost (your config file have localhost in hosts: section)?

**Reply by Saphar** [reply]
It is not, I just saw that in different discussions.

In the hosts I have the domain name (I will not type the exact domain here, but it looks like this):
hosts:
  - "xmpp.domain.net"

As I said the command:
`ejabberdctl export2sql xmpp.domain.net /tmp/mnesia.sql`
fails with RCP connectin failure:
`Failed RPC connection to the node ejabberd@localhost: timeout`

**Reply by prefiks** [reply]
Could you see if `ejabberdctl debug` works? If it works you could try executing this command from shell that this starts by running `ejd2sql:export("xmpp.domain.net", "/tmp/mnesia.sql").` (. at and is important).

**Reply by Saphar** [reply]
=CRITICAL REPORT==== 4-Aug-2023::05:57:20.359874 ===
Failed to start Erlang application 'p1_mysql': no such file or directory: p1_mysql.app. This usually means that ejabberd or Erlang was compiled/installed incorrectly.
ok

**Reply by prefiks** [reply]
Do you have sql server defined in your config? Do you have erlang-p1-mysql package installed?

**Reply by Saphar** [reply]
I did have:
sql_type: mysql
sql_server: "localhost"
sql_database: "ejabberd"
sql_username: "ejabberd"
sql_password: "password"

but not the default_db. I removed these lines and installed now the package and the message is different:
=ERROR REPORT==== 4-Aug-2023::06:21:26.194159 ===
Failed to start ejabberd_sql_sup: already_present
ok

We are getting there! 

Thank you for your help!

**Reply by prefiks** [reply]
Did you get after restarting server or when issuing that command?

**Reply by prefiks** [reply]
Also did you check if that command didn't generate sql file?

**Reply by Saphar** [reply]
I only reloaded the config, but I did a restart now and the command succeeds. Only that the exported file is empty.

Now I removed the file and tried again and the file does not appear.

**Reply by prefiks** [reply]
Did you get any output when running that command?

**Reply by Saphar** [reply]
yes, output was "ok"

**Reply by prefiks** [reply]
Could you see if running `ejd2sql:export(<<"xmpp.my.host">>, <<"/tmp/dump.sql">>, mod_mam).` works?

**Reply by prefiks** [reply]
Also if that works try `ejd2sql:export(<<"xmpp.my.host">>, <<"/tmp/dump.sql">>).`, looks like you need to provide strings that way to this command

**Reply by Saphar** [reply]
It runs with "ok", but still there is no file.

if I run: `su - ejabberd -c "touch /tmp/test"` it does put a file there, so permissions are good.

**Reply by Saphar** [reply]
Other question, if I skip this step will I be able to switch to SQL? (and just loose the archived messages?)

**Reply by prefiks** [reply, marked as answer]
That depends, if you want to switch all modules to use sql with data conversion, then all data will be not accessible (so also probably accounts, rosters, etc), but if you only set `db_type: sql` in mod_mam options, then only this module will not see old data, it will just write new data to sql db.

**Reply by Saphar** [reply]
Thank you. I did that, now it works with SQL.
